Blizzard.IO.RabbitMQ/NetqRabbitConsumer.cs

```diff
@@ -0,0 +1,100 @@
+    {
+        private readonly IBus _netqBus;
+        private readonly IDeserializer<TData> _deserializer;
+        private readonly IAbstractTypeDeserializer<TData> _abstractDeserializer;
+        private readonly IQueue _netqQueue;
+        private readonly IConverter<MessageProperties, RabbitMessageProperties> _messagePropertiesToRabbitMessagePropertiesConverter;
+        private readonly ILogger _logger;
+        
+        private IDisposable _consumeHandler;
+        private readonly Func<byte[], MessageProperties, TData> _dataExtractionStrategy;
+
+        public event Action<TData> MessageReceived;
+        public event Action<TData, RabbitMessageProperties> MessageWithMetadataReceived;
+
+        protected NetqRabbitConsumer(IBus netqBus, RabbitQueue sourceQueue, ILoggerFactory loggerFactory,
+            IConverter<MessageProperties, RabbitMessageProperties> messagePropertiesToRabbitMessagePropertiesConverter)
+        {
+            _netqBus = netqBus;
+            _netqQueue = new Queue(sourceQueue.Name, sourceQueue.Exclusive);
+            _logger = loggerFactory.CreateLogger(nameof(NetqRabbitConsumer<TData>));
+            _messagePropertiesToRabbitMessagePropertiesConverter = messagePropertiesToRabbitMessagePropertiesConverter ?? new RabbitPropertiesConverter();
+        }
+
+        public NetqRabbitConsumer(IBus netqBus, RabbitQueue sourceQueue, IDeserializer<TData> deserializer, ILoggerFactory loggerFactory,
+            IConverter<MessageProperties, RabbitMessageProperties> messagePropertiesToRabbitMessagePropertiesConverter = null)
+            : this(netqBus, sourceQueue, loggerFactory, messagePropertiesToRabbitMessagePropertiesConverter)
+        {
+            _deserializer = deserializer;
+            _dataExtractionStrategy += ExtractConcreteData;
+            _logger.LogDebug($"Initialized consumer from queue: {_netqQueue.Name}," +
+                $" expecting {typeof(TData)} data");
+        }
+
+        public NetqRabbitConsumer(IBus netqBus, RabbitQueue sourceQueue, IAbstractTypeDeserializer<TData> abstractDeserializer,
+            ILoggerFactory loggerFactory,
+            IConverter<MessageProperties, RabbitMessageProperties> messagePropertiesToRabbitMessagePropertiesConverter = null)
+            : this(netqBus, sourceQueue, loggerFactory, messagePropertiesToRabbitMessagePropertiesConverter)
+        {
+            _abstractDeserializer = abstractDeserializer;
+            _dataExtractionStrategy += ExtractAbstractData;
+            _logger.LogDebug($"Initialized consumer from queue: {_netqQueue.Name}," +
+                $" expecting {typeof(TData)} derieved data");
+        }
+
+        public void Start()
+        {
+            if (MessageReceived == null && MessageWithMetadataReceived == null)
+            {
+                _logger.LogWarning("Couldn't start consumer, must have at least one subscriber in order to start");
```

Comment:
Throw an exception dont write log.
We need to show the user his start function has failed I think that log is not good enough